### ISSUE TITLE
Setting environment variables to non-constant expressions

### ISSUE THREAD

**Issue by beatngu13** [junit-pioneer/junit-pioneer#381]
Recap #369:

> Annotation values must be constant expressions. That is:
> 
> ```java
> @ClearSystemProperty(key = "must be a constant expression")
> ```
> 
> A user may have to set a system property to a value that is not a constant expression. In that case, they can still leverage the restore mechanism of our system property extension:
> 
> ```java
> @ParameterizedTest
> @ValueSource(strings = { "foo", "bar" })
> @ClearSystemProperty(key = "some property")
> void test(String value) {
> 	System.setProperty("some property", value);
> }
> ```

So while there is a workaround for this in the case of system properties, there is none (to my knowledge) for environment variables. A use case for this (see [Stack Overflow](https://stackoverflow.com/a/64892391/3429133)):

>The https://junit-pioneer.org/ alternative requires environment variable values to be known at compile time. The above [using [System Stubs for JUnit Jupiter](https://github.com/webcompere/system-stubs/blob/master/system-stubs-jupiter/README.md)] also supports the setting of environment variables in the `@BeforeAll` [we do this too], which means it interoperates well with things like `Testcontainers` that might set up some resources needed by child tests.

To support this, we could provide something like:

```java
@ParameterizedTest
@ValueSource(strings = { "foo", "bar" })
@ClearEnvironmentVariable(key = "some property")
void test(String value) {
	EnvironmentVariable.set("some environment variable", value);
}
```

Internally, we currently do this using the non-public class [`EnvironmentVariableUtils`](https://github.com/junit-pioneer/junit-pioneer/blob/master/src/main/java/org/junitpioneer/jupiter/EnvironmentVariableUtils.java).

What do you think?

**Comment by nipafx** [comment]
I've read #450 and this one. I think we should do more than offer an imperative way to set these variables (via an internal API) and then reset them with `@Clear...`. A solution that works for both system properties and environment variables would be cool.

I liked your idea to solve this by referencing a method. We could add a `methodName` field (exact name t.b.d.) to the `@Set...` annotations and call that method to compute the value.

**Comment by beatngu13** [comment]
@nipafx I agree that adding simple setters is not so nice and also doesn't fit well with our overall API design. However, a `methodName` is probably not sufficient. In contrast to e.g. `@MethodSource`s, these methods are not necessarily only suppliers, functions might be needed, too.

According to the Stack Overflow post mentioned above, Testcontainers would be an example where one has to set environment variables and/or system properties at runtime. Unfortunately, I haven't used Testcontainers yet (😕), but maybe a team member such as @kiview or @bsideup finds time to give us a little example where and how non-constant expressions are needed? At least this would make designing the API easier. (I assume something like: container spins up, expose used port via an environment variable or system property to the class under test?)

**Comment by bsideup** [comment]
@beatngu13 TBH I haven't seen a use case where Testcontainers-provided values needed to be set as environment variables and cannot be provided as system properties.

That said, I would very much appreciate seeing `EnvironmentVariable.set`, as I needed it many times myself :)

I hope that helps!

**Comment by kiview** [comment]
To give a concrete example, a user might need to inject the result of `getJdbcUrl()` into their SUT after the container has started. However, the way by which the value needs to be injected is specific to their SUT or framework and independent from using Testcontainers (and outside of the responsibilities of Testcontainers).

**Comment by beatngu13** [comment]
Thank you both for getting back to us! ✌️

**Comment by nipafx** [comment]
I agree that referencing a method may be too limiting and it's pretty boilerplatey. Also, it would be nice to bring our resetting machinery to the table. What about this?

* provide arguments for parameters of type `SystemProperties` and `EnvironmentVariables`
* they can be used to clear and set properties/variables
* on clear and set, they store the initial value and reset it after the method

That would also alleviate the need for `@Clear...` on top of the method. Although we still need some way to apply the extensions...

**Comment by beatngu13** [comment]
[System Rules](https://stefanbirkner.github.io/system-rules/#RestoreSystemProperties) has `RestoreSystemProperties`. We could do something similar:

```java
@ResetSystemProperties
void testSysProps(SystemProperties sysProps) {
    sysProps.set("key", "value");
    // ...
}

@ResetEnvironmentVariables
void testEnvVars(EnvironmentVariables envVars) {
    envVars.set("key", "value");
    // ...
}
```

"Reset…", "Restore…", "Backup…", … ? In any way, it should be clear what the extension is all about.

**Comment by Marcono1234** [comment]
> [System Rules](https://stefanbirkner.github.io/system-rules/#RestoreSystemProperties) has `RestoreSystemProperties`. We could do something similar (using [`TypeBasedParameterResolver`](https://junit.org/junit5/docs/5.8.2/api/org.junit.jupiter.api/org/junit/jupiter/api/extension/support/TypeBasedParameterResolver.html))

Is the `SystemProperties` parameter really needed? Wouldn't it be easier and cover more cases, e.g. when the tested code modifies system properties, to cover regular `System.setProperty` usage?

Currently this issue is mainly about environment variables; would it be ok if I created a new issue for a `@RestoreSystemProperties` annotation (should also be decided whether that should include the `@BeforeEach` and `@AfterEach` actions). The developers of Log4j 2 seem to be looking for such an test extension, see https://github.com/apache/logging-log4j2/pull/658#issuecomment-1001159235.

**Comment by beatngu13** [comment]
> Is the `SystemProperties` parameter really needed? Wouldn't it be easier and cover more cases, e.g. when the tested code modifies system properties, to cover regular `System.setProperty` usage?

If we do create a backup of all system properties, the parameter is basically not needed. However, it is needed for environment variables as there is no `System#setenv(String, String)`. So I think it is more or less a matter of taste, if we want to keep the API design of both extensions similar.

> Currently this issue is mainly about environment variables; would it be ok if I created a new issue for a `@RestoreSystemProperties` annotation (should also be decided whether that should include the `@BeforeEach` and `@AfterEach` actions). The developers of Log4j 2 might be looking for such a test extension, see [apache/logging-log4j2#658 (comment)](https://github.com/apache/logging-log4j2/pull/658#issuecomment-1001159235).

Thank you for the Log4j issue link, interesting discussion! And yes, you could create a separate issue for resetting all system properties (although that is thightly coupled to the issue here).

Regarding `@BeforeEach` and `@AfterEach`: this relates to #479.

**Comment by stefanroeck** [comment]
Another use case for this is storing a filepath in an environment variable which is calculated during test setup based on a temp directory (e.g. using [JUnit @TempDir](https://junit.org/junit5/docs/5.9.2/api/org.junit.jupiter.api/org/junit/jupiter/api/io/TempDir.html))

```java
@BeforeEach
void setup(@TempDir Path tempDir) {
  this.tempDir = tempDir
}

@Test
@RestoreEnvironmentVariables
void testWithPathFromEnv() {
  // TODO: Programmatically set environment variable here based on this.tempDir
}
```

**Comment by nielsbasjes** [comment]
For reference my use case.
I have a component that is to be used in a Gitlab CI pipeline and needs to connect to the Gitlab instance to use the Gitlab APIs.

I use `@WireMockTest` (see [here](https://github.com/nielsbasjes/codeowners/blob/e1c3f694b3027ebc31e83c10dedaef5561cb91dc/enforcer/src/test/java/nl/basjes/maven/enforcer/codeowners/TestGitlabUsers.java#L37)) to mock the Gitlab server behavior.
The URL of the Gitlab server (here mocked) needs to be set in the environment variable `CI_SERVER_URL` (as defined [here](https://docs.gitlab.com/ci/variables/predefined_variables/)) for my software to think it is in the real environment.

**Comment by nielsbasjes** [comment]
Thanks for @Michael1993 pointing me to this issue which is labeled as `up for grabs`.

So I made a first attempt to implement this: https://github.com/junit-pioneer/junit-pioneer/pull/872

This is a very draft pull request (code is not clean, not enough tests, bad documentation, ... ).

The key question right now is if this is an implementation direction you like (Or did I completely get it wrong?).

I went with the approach of making new annotations that are `FromSource`.
I needed to store the indexes of the desired parameters in a Store and then later during the interceptTestTemplateMethod do the actual setting of the values.

The main thing is that it passes the test I created.
(Updated: I had forgotten to copy the essential annotations)
```
@ParameterizedTest 
 @MethodSource("provideParameters")
	@SetEnvironmentVariableFromSource(key = "set envvar LEFT", parameter = 0)
	@SetEnvironmentVariableFromSource(key = "set envvar RIGHT", parameter = 1)
	public void testParametersFromMethod(String name, int value) {
		assertThat(systemEnvironmentVariable("set envvar LEFT")).isEqualTo(name);
		assertThat(systemEnvironmentVariable("set envvar RIGHT")).isEqualTo(""+value);
	}

	private static Stream<Arguments> provideParameters() {
		return Stream.of(
				Arguments.of("one", 1),
				Arguments.of("two", 2),
				Arguments.of("three", 3)
		);
```

**Comment by nielsbasjes** [comment]
I cleaned up my changes, wrote tests and documentation.
From where I'm standing this looks quite good.

With this you can now do this (and similar for System Properties) :
```java
@ParameterizedTest(name = "Setting MyEnvVariable to {0}")
@ValueSource(strings = { "foo", "bar" })
@SetEnvironmentVariableFromSource(key = "MyEnvVariable", argument = 0)
@SetEnvironmentVariable(key = "MyOtherEnvVariable", value = "Something")
@RestoreEnvironmentVariables
void parameterizedFromSourceTest(String value) {
    assertThat(System.getenv("MyEnvVariable")).isEqualTo(value);
    assertThat(System.getenv("MyOtherEnvVariable")).isEqualTo("Something");
}
```

The `@RestoreEnvironmentVariables` is needed because of #875 

Since it is based on the JUnit `Source` system you can apply any kind of `Source` that is suitable for the specific use case. I have done tests with ValueSource, MethodSource and FieldSource and all work fine. The only minor downside is that the environment variable value must be an argument to the test method also. This however makes it easy to double check if the environment has been set correctly as shown in the above example.

@beatngu13 / @Michael1993 I would really appreciate feedback on this proposed implementation.
